vertical-pod-autoscaler/pkg/admission-controller/resource/workloads/utils.go

```diff
@@ -0,0 +1,120 @@
+	vpaMatcher       vpa.Matcher
+	patchCalculators []patch.Calculator
+}
+
+// NewResourceHandler creates new instance of resourceHandler.
+func NewResourceHandler(preProcessor pod.PreProcessor, vpaMatcher vpa.Matcher, patchCalculators []patch.Calculator) *WorkloadResourceHandler {
+	return &WorkloadResourceHandler{
+		preProcessor:     preProcessor,
+		vpaMatcher:       vpaMatcher,
+		patchCalculators: patchCalculators,
+	}
+}
+
+// GetPatchesForWorkload builds patches for Pod in given admission request.
+func (h *WorkloadResourceHandler) GetPatchesForWorkload(namespace string, objectMeta *metav1.ObjectMeta, podTemplateSpec *v1.PodTemplateSpec) ([]resource_admission.PatchRecord, error) {
+	if len(objectMeta.Name) == 0 {
+		objectMeta.Name = objectMeta.GenerateName + "%"
+		objectMeta.Namespace = namespace
+	}
+
+	// We create a `Pod` to be able to re-use the pod code.
+	pod := PodFromPodTemplateSpec(objectMeta, podTemplateSpec)
+	klog.V(4).Infof("Generating pod %s/%s", pod.Namespace, pod.Name)
+	controllingVpa := h.vpaMatcher.GetMatchingVPA(&pod)
+	if controllingVpa == nil {
+		klog.V(4).Infof("No matching VPA found for %s/%s", objectMeta.Namespace, objectMeta.Name)
+		return []resource_admission.PatchRecord{}, nil
+	}
+	pod, err := h.preProcessor.Process(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	patches := []resource_admission.PatchRecord{}
+	if objectMeta.Annotations == nil {
+		patches = append(patches, patch.GetAddEmptyAnnotationsPatch())
+	}
+	for _, c := range h.patchCalculators {
+		partialPatches, err := c.CalculatePatches(&pod, controllingVpa)
+		if err != nil {
+			return []resource_admission.PatchRecord{}, err
+		}
+		patches = append(patches, partialPatches...)
+	}
+
+	return FixupPodPatchesForWorkload(patches), nil
+}
+
+// FixupPodPatchesForWorkload takes patches for a pod and adapt them to be applied to a workload.
+func FixupPodPatchesForWorkload(patches []resource_admission.PatchRecord) []resource_admission.PatchRecord {
+	for i, patch := range patches {
+		// If it's a patch to spec, add the path to the PodSpec
+		if strings.HasPrefix(patch.Path, "/spec/") {
+			patches[i].Path = "/spec/template" + patch.Path
+		}
+	}
+	return patches
+}
+
+// PodFromPodTemplateSpec creates a fake Pod object from a PodTemplateSpec. This is used by the workload
+// admission controllers tu re-use as much code as we can without modifying it, it makes rebases easier.
+// If we ever want to upstream this code we should make the pod code more generic.
+func PodFromPodTemplateSpec(objectMeta *metav1.ObjectMeta, podTemplateSpec *v1.PodTemplateSpec) v1.Pod {
+	pod := v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		// This is important to get the same labels as the pod (which need to match the selector)
+		ObjectMeta: podTemplateSpec.ObjectMeta,
+		Spec:       podTemplateSpec.Spec,
+		Status:     v1.PodStatus{},
+	}
+	// We also need to propagate the trigger annotation that was likely set on the parent.
```

Comment:
I don't understand why we NEED to do that. This annotation should be set on the deployment and the STS only, what role does it play at pod level ?

Reply:
We try to re-use most of the pod code, and the pod code will get a vpa that is not disabled, and if this vpa has UpdateMode=Trigger it will want the annotation to be on the "pod". But maybe I can rework this to not work on single pods and do the trigger check only for workloads.

Reply:
Fine I got it in the end.